pkg/remote/remote.go

```diff
@@ -0,0 +1,320 @@
+			}
+			defer res.Body.Close()
+			b, err := ioutil.ReadAll(res.Body)
+			if err != nil {
+				log.Infof("Request not found for remote doctype %s: %s", doctype, err)
+				return nil, ErrNotFoundRemote
+			}
+			dt.Request = string(b)
+			err = couchdb.CreateNamedDocWithDB(ins, &dt)
+			if err != nil {
+				log.Infof("Cannot save remote doctype %s: %s", doctype, err)
+			}
+		}
+		raw = dt.Request
+	} else {
+		filename := path.Join(config.GetConfig().Doctypes, doctype, "request")
+		bytes, err := ioutil.ReadFile(filename)
+		if err != nil {
+			return nil, ErrNotFoundRemote
+		}
+		raw = string(bytes)
+	}
+
+	return ParseRawRequest(doctype, raw)
+}
+
+// ExtractVariables extracts the variables:
+// - from the query string for a GET
+// - from the body formatted as JSON for a POST
+func ExtractVariables(verb string, in *http.Request) (map[string]string, error) {
+	vars := make(map[string]string)
+	if verb == echo.GET {
+		for k, v := range in.URL.Query() {
+			vars[k] = v[0]
+		}
+	} else {
+		err := json.NewDecoder(in.Body).Decode(&vars)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return vars, nil
+}
+
+var injectionRegexp = regexp.MustCompile(`{{\w+}}`)
+
+func injectVar(src string, vars map[string]string) string {
+	return injectionRegexp.ReplaceAllStringFunc(src, func(m string) string {
+		m = strings.TrimLeft(m, "{{")
+		m = strings.TrimRight(m, "}}")
+		m = strings.TrimSpace(m)
+		return vars[m]
+	})
+}
+
+// InjectVariables replaces {{variable}} by its value in some fields of the
+// remote struct
+func InjectVariables(remote *Remote, vars map[string]string) {
+	if strings.Contains(remote.URL.Path, "{{") {
+		remote.URL.Path = injectVar(remote.URL.Path, vars)
+	}
+	if strings.Contains(remote.URL.RawQuery, "{{") {
+		remote.URL.RawQuery = injectVar(remote.URL.RawQuery, vars)
+	}
+	for k, v := range remote.Headers {
+		if strings.Contains(v, "{{") {
+			remote.Headers[k] = injectVar(v, vars)
+		}
+	}
+	if strings.Contains(remote.Body, "{{") {
+		remote.Body = injectVar(remote.Body, vars)
+	}
+}
+
+// ProxyTo calls the external website and proxy the reponse
+func (remote *Remote) ProxyTo(doctype string, ins *instance.Instance, rw http.ResponseWriter, in *http.Request) error {
+	vars, err := ExtractVariables(remote.Verb, in)
+	if err != nil {
+		log.Infof("Error on extracting variables: %s", err)
+		return ErrInvalidVariables
+	}
+	InjectVariables(remote, vars)
+
+	// Sanitize the remote URL
+	if strings.Contains(remote.URL.Host, ":") {
+		log.Infof("Invalid host for remote doctype %s: %s", doctype, remote.URL.Host)
+		return ErrInvalidRequest
+	}
+	remote.URL.User = nil
+	remote.URL.Fragment = ""
+
+	req, err := http.NewRequest(remote.Verb, remote.URL.String(), nil)
+	if err != nil {
+		return ErrInvalidRequest
+	}
+
+	req.Header.Set("User-Agent", "cozy-stack "+config.Version+" ("+runtime.Version()+")")
+	for k, v := range remote.Headers {
+		req.Header.Set(k, v)
+	}
+
+	res, err := http.DefaultClient.Do(req)
```

Comment:
I think the default client has a infinite timeout which could be problematic in the case of a proxy, with stall connection blocking others. The use of a specific `http.Client` instance with more sensible defaults may be useful here.